### PR TITLE
Added error decoding to roboclaw_wrapper

### PR DESCRIPTION
**Pull request by abust005**
Added error decoding to roboclaw_wrapper, using the error code table in the Roboclaw manual.

Also statically typed the parameters, as I encountered some issues running the code with the dynamic typing.

### CODE REVIEW

File: ROS/osr_control/osr_control/roboclaw_wrapper.py [hunk 1]
```diff
@@ -420,16 +421,94 @@ def stop_motors(self):
 
     def read_errors(self):
         """Checks error status of each motor controller, returns 0 if no errors reported"""
-        err = [0] * 3
+        err = ['0'] * 5
```
Comment (Achllle) [inline ROS/osr_control/osr_control/roboclaw_wrapper.py:424]: error status has [length 3](https://github.com/nasa-jpl/osr-rover-code/blob/37b6033e2b718f5a65e54fffdc9d0a01897720c3/ROS/osr_interfaces/msg/Status.msg#L4)

```suggestion
        err = ['0'] * 3
```

File: ROS/osr_control/osr_control/roboclaw_wrapper.py [hunk 2]
```diff
@@ -420,16 +421,94 @@ def stop_motors(self):
...
-            # convert to hexadecimal and then to string for easy decoding
-            err[i] = str(hex(err_int))
+
             if err_int != 0:
-                self.log.error(f"Motor controller '{self.address[i]}' reported error code {err[i]}")
-        
+                # convert to hexadecimal and then to string for easy decoding
+                err[i] = str(hex(err_int))
+
+                errString, hasError = self.decode_error(err_int)
```
Comment (Achllle) [inline ROS/osr_control/osr_control/roboclaw_wrapper.py:432]:
```suggestion
                err_string, has_error = self.decode_error(err_int)
```

Follow [PEP8](https://peps.python.org/pep-0008/#function-and-variable-names)

**Review by Achllle — CHANGES_REQUESTED** [review]
Nice, would like to get this improvement in.

Some smaller suggestions, PEP8.

> Also statically typed the parameters, as I encountered some issues running the code with the dynamic typing.

Can you clarify what kind of issues? For Python, this shouldn't matter, additionally, having them undefined allows you to accept different types. For example if you specify some parameter needs to be a `double` and someone writes `5` instead of `5.0`, it will reject that value.

File: ROS/osr_control/osr_control/roboclaw_wrapper.py (hunk 1 above)
Comment (abust005) [reply ROS/osr_control/osr_control/roboclaw_wrapper.py]: See commit ede2052

File: ROS/osr_control/osr_control/roboclaw_wrapper.py (hunk 2 above)
Comment (abust005) [reply ROS/osr_control/osr_control/roboclaw_wrapper.py]: See commit ede2052

**Comment by abust005** [comment]
> Can you clarify what kind of issues? 

For some reason, when I tried to use the dynamic typing functionality for the parameters, the scripts would actually crash on startup. I would get an InvalidParameterTypeException. See below:

`[roboclaw_wrapper-1] Traceback (most recent call last):
[roboclaw_wrapper-1]   File "/home/bobert/osr_ws/install/osr_control/lib/osr_control/roboclaw_wrapper", line 33, in <module>
[roboclaw_wrapper-1]     sys.exit(load_entry_point('osr-control', 'console_scripts', 'roboclaw_wrapper')())
[roboclaw_wrapper-1]   File "/home/bobert/osr_ws/build/osr_control/osr_control/roboclaw_wrapper.py", line 437, in main
[roboclaw_wrapper-1]     wrapper = RoboclawWrapper()
[roboclaw_wrapper-1]   File "/home/bobert/osr_ws/build/osr_control/osr_control/roboclaw_wrapper.py", line 28, in __init__
[roboclaw_wrapper-1]     self.declare_parameters(
[roboclaw_wrapper-1]   File "/opt/ros/iron/lib/python3.10/site-packages/rclpy/node.py", line 526, in declare_parameters
[roboclaw_wrapper-1]     self._declare_parameter_common(
[roboclaw_wrapper-1]   File "/opt/ros/iron/lib/python3.10/site-packages/rclpy/node.py", line 582, in _declare_parameter_common
[roboclaw_wrapper-1]     raise InvalidParameterTypeException(
[roboclaw_wrapper-1] rclpy.exceptions.InvalidParameterTypeException: Trying to set parameter 'baud_rate' to '115200' of type 'INTEGER', expecting type 'NOT_SET'
[ERROR] [roboclaw_wrapper-1]: process has died [pid 1612, exit code 1, cmd '/home/bobert/osr_ws/install/osr_control/lib/osr_control/roboclaw_wrapper --ros-args -r __node:=roboclaw_wrapper --params-file /home/bobert/osr_ws/install/osr_bringup/share/osr_bringup/config/roboclaw_params.yaml'].
[rover-3] [INFO] [1697965140.669282010] [rover]: Initializing Rover
[rover-3] Traceback (most recent call last):
[rover-3]   File "/home/bobert/osr_ws/install/osr_control/lib/osr_control/rover", line 33, in <module>
[rover-3]     sys.exit(load_entry_point('osr-control', 'console_scripts', 'rover')())
[rover-3]   File "/home/bobert/osr_ws/build/osr_control/osr_control/rover.py", line 400, in main
[rover-3]     rover = Rover()
[rover-3]   File "/home/bobert/osr_ws/build/osr_control/osr_control/rover.py", line 24, in __init__
[rover-3]     self.declare_parameters(
[rover-3]   File "/opt/ros/iron/lib/python3.10/site-packages/rclpy/node.py", line 526, in declare_parameters
[rover-3]     self._declare_parameter_common(
[rover-3]   File "/opt/ros/iron/lib/python3.10/site-packages/rclpy/node.py", line 582, in _declare_parameter_common
[rover-3]     raise InvalidParameterTypeException(
[rover-3] rclpy.exceptions.InvalidParameterTypeException: Trying to set parameter 'rover_dimensions.d1' to '0.177' of type 'DOUBLE', expecting type 'NOT_SET'
[ERROR] [rover-3]: process has died [pid 1616, exit code 1, cmd '/home/bobert/osr_ws/install/osr_control/lib/osr_control/rover --ros-args -r __node:=rover --params-file /home/bobert/osr_ws/install/osr_bringup/share/osr_bringup/config/osr_params.yaml --params-file /tmp/launch_params_v59momyz'].`

File: ROS/osr_control/osr_control/roboclaw_wrapper.py (hunk 2 above)
Comment (abust005) [reply ROS/osr_control/osr_control/roboclaw_wrapper.py]: And commit e1cdd66

**Review by Achllle — APPROVED** [review]
Amazing, this will go a long way for helping people debug what's going on.

**Comment by Achllle** [comment]
> For some reason, when I tried to use the dynamic typing functionality for the parameters, the scripts would actually crash on startup. I would get an InvalidParameterTypeException.

Perhaps this is because I'm on ROS2 Foxy still, haven't yet updated and behavior changed with Humble/Iron. 

Let me know if you think this is ready and we'll get it merged in!

**Comment by abust005** [comment]
Please perform one last review to make sure I've pulled all the parameter static typing; if it's clean, then we're good to merge!

File: ROS/osr_control/osr_control/roboclaw_wrapper.py [hunk 3]
```diff
@@ -59,7 +59,7 @@ def __init__(self):
                 ('roboclaw_mapping.drive_right_back.channel', None),
                 ('roboclaw_mapping.drive_right_back.ticks_per_rev', None),
                 ('roboclaw_mapping.drive_right_back.gear_ratio', None)
-            ]
+             ]
```
Comment (Achllle) [inline ROS/osr_control/osr_control/roboclaw_wrapper.py:62]:
```suggestion
            ]
```
this seems to be a typo, if anything this should be 4 spaces back. Now it seems to be 5 spaces in.

File: ROS/osr_control/osr_control/rover.py [hunk 4]
```diff
@@ -24,10 +24,10 @@ def __init__(self):
         self.declare_parameters(
             namespace='',
             parameters=[
-                ('rover_dimensions.d1', None),
                 ('rover_dimensions.d2', None),
                 ('rover_dimensions.d3', None),
                 ('rover_dimensions.d4', None),
+                ('rover_dimensions.d1', None),
```
Comment (Achllle) [inline ROS/osr_control/osr_control/rover.py:30]: Seems like you flipped order here, I think 1,2,3,4 made more sense :)

File: ROS/osr_control/osr_control/roboclaw_wrapper.py (hunk 3 above)
Comment (abust005) [reply ROS/osr_control/osr_control/roboclaw_wrapper.py]: Good catch! Commit 7bbf40e :)

File: ROS/osr_control/osr_control/rover.py (hunk 4 above)
Comment (abust005) [reply ROS/osr_control/osr_control/rover.py]: Absolutely! Didn't even notice I did that; fixed in commit 7bbf40e

**Review by Achllle — APPROVED** [review]
Great work, thanks for making the updates!

**Pull request merged**